### ISSUE TITLE
`fail_when_catalog_not_accessible` should not be used during plan

### ISSUE THREAD

**Issue by mogul** [cloudfoundry-community/terraform-provider-cloudfoundry#292]
From #269 :
> I've closed the issue but if you find or think there is a biggest problem feel free to reopen

I'm unable to re-open so forgive me not just appending there.

I deleted a broker app, then deleted the broker registration(s) with CF, then ran terraform to repopulate them. Initially I got:

```
cloudfoundry_service_broker.space-scoped-broker["gsa-datagov/development"]: Refreshing state... [id=17d997ce-4e09-47bf-ace8-6ab284a8dd5d]
cloudfoundry_service_broker.space-scoped-broker["gsa-datagov/staging"]: Refreshing state... [id=2920897c-739f-4999-bbcc-4954dbc38230]
2020/11/20 21:41:56 [TRACE] <root>: eval: *terraform.EvalRefreshDependencies
2020/11/20 21:41:56 [TRACE] <root>: eval: *terraform.EvalRefresh
2020/11/20 21:41:56 [TRACE] GRPCProvider: ReadResource
cloudfoundry_service_broker.space-scoped-broker["gsa-datagov/prod"]: Refreshing state... [id=9d900e6d-d81a-4dbf-885e-020f098a75be]
Error: /20 21:41:57 [ERROR] <root>: eval: *terraform.EvalRefresh, err: Error when getting catalog signature: Status code: 404 Not Found, Body: 404 Not Found: Requested route ('ssb-liberal-jackass.app.cloud.gov') does not exist.
Error: /20 21:41:57 [ERROR] <root>: eval: *terraform.EvalSequence, err: Error when getting catalog signature: Status code: 404 Not Found, Body: 404 Not Found: Requested route ('ssb-liberal-jackass.app.cloud.gov') does not exist.
2020/11/20 21:41:57 [TRACE] [walkRefresh] Exiting eval tree: cloudfoundry_service_broker.space-scoped-broker["gsa-datagov/development"]
2020/11/20 21:41:57 [TRACE] vertex "cloudfoundry_service_broker.space-scoped-broker[\"gsa-datagov/development\"]": visit complete
Error: /20 21:41:57 [ERROR] <root>: eval: *terraform.EvalRefresh, err: Error when getting catalog signature: Status code: 404 Not Found, Body: 404 Not Found: Requested route ('ssb-liberal-jackass.app.cloud.gov') does not exist.
Error: /20 21:41:57 [ERROR] <root>: eval: *terraform.EvalRefresh, err: Error when getting catalog signature: Status code: 404 Not Found, Body: 404 Not Found: Requested route ('ssb-liberal-jackass.app.cloud.gov') does not exist.
Error: /20 21:41:57 [ERROR] <root>: eval: *terraform.EvalSequence, err: Error when getting catalog signature: Status code: 404 Not Found, Body: 404 Not Found: Requested route ('ssb-liberal-jackass.app.cloud.gov') does not exist.
2020/11/20 21:41:57 [TRACE] [walkRefresh] Exiting eval tree: cloudfoundry_service_broker.space-scoped-broker["gsa-datagov/staging"]
2020/11/20 21:41:57 [TRACE] vertex "cloudfoundry_service_broker.space-scoped-broker[\"gsa-datagov/staging\"]": visit complete
Error: /20 21:41:57 [ERROR] <root>: eval: *terraform.EvalSequence, err: Error when getting catalog signature: Status code: 404 Not Found, Body: 404 Not Found: Requested route ('ssb-liberal-jackass.app.cloud.gov') does not exist.
 
2020/11/20 21:41:57 [TRACE] [walkRefresh] Exiting eval tree: cloudfoundry_service_broker.space-scoped-broker["gsa-datagov/prod"]
2020/11/20 21:41:57 [TRACE] vertex "cloudfoundry_service_broker.space-scoped-broker[\"gsa-datagov/prod\"]": visit complete
2020/11/20 21:41:57 [TRACE] dag/walk: upstream of "root" errored, so skipping
2020/11/20 21:41:57 [TRACE] vertex "cloudfoundry_service_broker.space-scoped-broker": dynamic subgraph encountered errors
2020/11/20 21:41:57 [TRACE] vertex "cloudfoundry_service_broker.space-scoped-broker": visit complete
2020/11/20 21:41:57 [TRACE] dag/walk: upstream of "provider.cloudfoundry (close)" errored, so skipping
2020/11/20 21:41:57 [TRACE] dag/walk: upstream of "root" errored, so skipping
Error: Error when getting catalog signature: Status code: 404 Not Found, Body: 404 Not Found: Requested route ('ssb-liberal-jackass.app.cloud.gov') does not exist.
Error: Error when getting catalog signature: Status code: 404 Not Found, Body: 404 Not Found: Requested route ('ssb-liberal-jackass.app.cloud.gov') does not exist.
Error: Error when getting catalog signature: Status code: 404 Not Found, Body: 404 Not Found: Requested route ('ssb-liberal-jackass.app.cloud.gov') does not exist.
2020-11-20T21:41:57.346Z [WARN]  plugin.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = transport is closing"
2020-11-20T21:41:57.347Z [DEBUG] plugin: plugin process exited: path=/root/.terraform.d/plugins/linux_amd64/terraform-provider-cloudfoundry_v0.12.3 pid=70
2020-11-20T21:41:57.347Z [DEBUG] plugin: plugin exited
Error: Process completed with exit code 1.
```

OK, so referring back to #529, I should be able to set `fail_when_catalog_not_accessible` to `false` to get past that situation, right? However, **even when I set the parameter to `false` and re-ran the terraform**, I got this in the trace:
```
cloudfoundry_service_broker.space-scoped-broker["gsa-datagov/prod"]: Refreshing state... [id=9d900e6d-d81a-4dbf-885e-020f098a75be]
cloudfoundry_service_broker.space-scoped-broker["gsa-datagov/development"]: Refreshing state... [id=17d997ce-4e09-47bf-ace8-6ab284a8dd5d]
cloudfoundry_service_broker.space-scoped-broker["gsa-datagov/staging"]: Refreshing state... [id=2920897c-739f-4999-bbcc-4954dbc38230]
Error: /20 21:44:40 [ERROR] <root>: eval: *terraform.EvalRefresh, err: Error when getting catalog signature: Status code: 404 Not Found, Body: 404 Not Found: Requested route ('ssb-liberal-jackass.app.cloud.gov') does not exist.
 
Error: /20 21:44:40 [ERROR] <root>: eval: *terraform.EvalSequence, err: Error when getting catalog signature: Status code: 404 Not Found, Body: 404 Not Found: Requested route ('ssb-liberal-jackass.app.cloud.gov') does not exist.
 
2020/11/20 21:44:40 [TRACE] [walkRefresh] Exiting eval tree: cloudfoundry_service_broker.space-scoped-broker["gsa-datagov/staging"]
2020/11/20 21:44:40 [TRACE] vertex "cloudfoundry_service_broker.space-scoped-broker[\"gsa-datagov/staging\"]": visit complete
Error: /20 21:44:40 [ERROR] <root>: eval: *terraform.EvalRefresh, err: Error when getting catalog signature: Status code: 404 Not Found, Body: 404 Not Found: Requested route ('ssb-liberal-jackass.app.cloud.gov') does not exist.
Error: /20 21:44:40 [ERROR] <root>: eval: *terraform.EvalSequence, err: Error when getting catalog signature: Status code: 404 Not Found, Body: 404 Not Found: Requested route ('ssb-liberal-jackass.app.cloud.gov') does not exist.
2020/11/20 21:44:40 [TRACE] [walkRefresh] Exiting eval tree: cloudfoundry_service_broker.space-scoped-broker["gsa-datagov/development"]
2020/11/20 21:44:40 [TRACE] vertex "cloudfoundry_service_broker.space-scoped-broker[\"gsa-datagov/development\"]": visit complete
Error: /20 21:44:41 [ERROR] <root>: eval: *terraform.EvalRefresh, err: Error when getting catalog signature: Status code: 404 Not Found, Body: 404 Not Found: Requested route ('ssb-liberal-jackass.app.cloud.gov') does not exist.
Error: /20 21:44:41 [ERROR] <root>: eval: *terraform.EvalSequence, err: Error when getting catalog signature: Status code: 404 Not Found, Body: 404 Not Found: Requested route ('ssb-liberal-jackass.app.cloud.gov') does not exist.
2020/11/20 21:44:41 [TRACE] [walkRefresh] Exiting eval tree: cloudfoundry_service_broker.space-scoped-broker["gsa-datagov/prod"]
2020/11/20 21:44:41 [TRACE] vertex "cloudfoundry_service_broker.space-scoped-broker[\"gsa-datagov/prod\"]": visit complete
2020/11/20 21:44:41 [TRACE] dag/walk: upstream of "root" errored, so skipping
2020/11/20 21:44:41 [TRACE] vertex "cloudfoundry_service_broker.space-scoped-broker": dynamic subgraph encountered errors
2020/11/20 21:44:41 [TRACE] vertex "cloudfoundry_service_broker.space-scoped-broker": visit complete
2020/11/20 21:44:41 [TRACE] dag/walk: upstream of "provider.cloudfoundry (close)" errored, so skipping
2020/11/20 21:44:41 [TRACE] dag/walk: upstream of "root" errored, so skipping
Error: Error when getting catalog signature: Status code: 404 Not Found, Body: 404 Not Found: Requested route ('ssb-liberal-jackass.app.cloud.gov') does not exist.
Error: Error when getting catalog signature: Status code: 404 Not Found, Body: 404 Not Found: Requested route ('ssb-liberal-jackass.app.cloud.gov') does not exist.
Error: Error when getting catalog signature: Status code: 404 Not Found, Body: 404 Not Found: Requested route ('ssb-liberal-jackass.app.cloud.gov') does not exist.
2020-11-20T21:44:41.016Z [WARN]  plugin.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = transport is closing"
2020-11-20T21:44:41.017Z [DEBUG] plugin: plugin process exited: path=/root/.terraform.d/plugins/linux_amd64/terraform-provider-cloudfoundry_v0.12.3 pid=85
2020-11-20T21:44:41.017Z [DEBUG] plugin: plugin exited
Error: Process completed with exit code 1.
```

In other words, `fail_when_catalog_not_accessible` seems to be ignored (and defaulting to `true`) during `terraform plan`.

**Comment by mogul** [comment]
I want to make it clear that this bug has no workaround. I tried adding "--refresh=no" on all of Terraform plan, apply, and destroy, and in all cases, the provider errored out. The *only* way to get out of this situation was to edit the Terraform state to remove all references to the broker, and then remove it by hand.

**Comment by ArthurHlt** [comment]
you should not have anymore this issue with the PR you can found above, thanks for the feedback and make things better

**Comment by mogul** [comment]
Thanks so much! This makes me feel much more confident in applying the broker for my use-case.
